### PR TITLE
Add sarif reporting for semgrep

### PR DESCRIPTION
**Pull request by Kamsiy**
## Sample json output 
```
{
  "version": "2.10.19",
  "project_name": "Neon Genesis",
  "passed": false,
  "scans": {
    "Semgrep": {
      "scanner_name": "Semgrep",
      "version": "0.42.0",
      "passed": false,
      "logs": "Could not parse unparsable_py.py as python (warn)\n\t/Users/kamsi.ibeziako/Documents/salus-main/spec/fixtures/semgrep/invalid/unparsable_py.py:3-3\n\nForbidden pattern \"$X == $X\" was found - Useless equality test.\n\texamples/trivial2.py:10:    if user.id == user.id:\n\nForbidden pattern \"$X == $X\" was found - Useless equality test.\n\ttrivial.py:3:if 3 == 3:\n\nForbidden pattern \"$X == $X\" was found - Useless equality test.\n\tvendor/trivial2.py:10:    if user.id == user.id:",
      "warn": {
        "semgrep_non_fatal": [
          {
            "type": "SourceParseError",
            "message": "Could not parse unparsable_py.py as python",
            "level": "warn",
            "spans": [
              {
                "file": "/Users/kamsi.ibeziako/Documents/salus-main/spec/fixtures/semgrep/invalid/unparsable_py.py",
                "start": {
                  "line": 3,
                  "col": 12
                },
                "end": {
                  "line": 3,
                  "col": 12
                }
              }
            ]
          }
        ]
      },
      "info": {
        "hits": [
          {
            "pattern": "$X == $X",
            "config": null,
            "forbidden": true,
            "required": false,
            "msg": "Useless equality test.",
            "hit": "examples/trivial2.py:10:    if user.id == user.id:"
          },
          {
            "pattern": "$X == $X",
            "config": null,
            "forbidden": true,
            "required": false,
            "msg": "Useless equality test.",
            "hit": "trivial.py:3:if 3 == 3:"
          },
          {
            "pattern": "$X == $X",
            "config": null,
            "forbidden": true,
            "required": false,
            "msg": "Useless equality test.",
            "hit": "vendor/trivial2.py:10:    if user.id == user.id:"
          }
        ],
        "misses": [

        ]
      },
      "errors": [

      ]
    }
  },
  "errors": [

  ]
}
```
## Sample SARIF output 
```
{
  "version": "2.1.0",
  "$schema": "https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/schemas/sarif-schema-2.1.0",
  "runs": [
    {
      "tool": {
        "driver": {
          "name": "Semgrep",
          "version": "0.42.0",
          "informationUri": "https://semgrep.dev/",
          "rules": [
            {
              "id": "Forbidden Pattern Found",
              "name": "$X == $X / Useless equality test. Forbidden Pattern Found",
              "fullDescription": {
                "text": "Pattern: $X == $X\nMessage:Useless equality test.\n\nForbidden:true"
              },
              "helpUri": "https://github.com/coinbase/salus/blob/master/docs/scanners/semgrep.md",
              "help": {
                "text": "More info: https://github.com/coinbase/salus/blob/master/docs/scanners/semgrep.md",
                "markdown": "[More info](https://github.com/coinbase/salus/blob/master/docs/scanners/semgrep.md)."
              }
            },
            {
              "id": "SourceParseError",
              "name": "SourceParseError",
              "fullDescription": {
                "text": "Could not parse unparsable_py.py as python"
              },
              "helpUri": "https://semgrep.dev/docs/writing-rules/rule-syntax/",
              "help": {
                "text": "More info: https://semgrep.dev/docs/writing-rules/rule-syntax/",
                "markdown": "[More info](https://semgrep.dev/docs/writing-rules/rule-syntax/)."
              }
            }
          ]
        }
      },
      "conversion": {
        "tool": {
          "driver": {
            "name": "Salus",
            "informationUri": "https://github.com/coinbase/salus"
          }
        }
      },
      "results": [
        {
          "ruleId": "Forbidden Pattern Found",
          "ruleIndex": 0,
          "level": "error",
          "message": {
            "text": "Pattern: $X == $X\nMessage:Useless equality test.\n\nForbidden:true"
          },
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "examples/trivial2.py",
                  "uriBaseId": "%SRCROOT%"
                },
                "region": {
                  "startLine": 10,
                  "startColumn": 1,
                  "snippet": {
                    "text": "    if user.id == user.id"
                  }
                }
              }
            }
          ]
        },
        {
          "ruleId": "SourceParseError",
          "ruleIndex": 1,
          "level": "error",
          "message": {
            "text": "Could not parse unparsable_py.py as python"
          },
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "/Users/kamsi.ibeziako/Documents/salus-main/spec/fixtures/semgrep/invalid/unparsable_py.py",
                  "uriBaseId": "%SRCROOT%"
                },
                "region": {
                  "startLine": 3,
                  "startColumn": 12
                }
              }
            }
          ]
        }
      ],
      "invocations": [
        {
          "executionSuccessful": false
        }
      ]
    }
  ]
}

### CODE REVIEW

File: spec/lib/sarif/semgrep_sarif_spec.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        report.add_scan_report(scanner.report, required: false)
+
+        sarif_report = JSON.parse(report.to_sarif)
+        result = sarif_report["runs"][0]["results"][0]
+        rules = sarif_report["runs"][0]["tool"]["driver"]["rules"]
+        runs_obj = sarif_report["runs"][0]
+        expect(rules[0]['id']).to eq("$X == $X")
+        expect(rules[0]['name']).to eq("$X == $X")
+        expect(runs_obj['invocations'][0]['executionSuccessful']).to eq(true)
+        expect(result['message']['text']).to eq("Pattern: $X == $X\nMessage:Useless equality test."\
```
Comment (joshua-ostrom) [inline spec/lib/sarif/semgrep_sarif_spec.rb:55]: This is failing on Circle FWIW
```
 -Hit: vendor/trivial2.py:10:    if user.id == user.id:
+Hit: examples/trivial2.py:10:    if user.id == user.id:
```

File: lib/sarif/semgrep_sarif.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      return nil if @issues.include?(id)
+
+      @issues.add(id)
+      location = hit[:hit].split(":") # [file_name, line, code_preview]
+      {
+        id: hit[:pattern],
+        name: "#{hit[:pattern]} / #{hit[:msg]}",
+        level: "HIGH",
+        details: "Pattern: #{hit[:pattern]}\nMessage:#{hit[:msg]}\nForbidden:#{hit[:forbidden]}"\
+        "\nRequired:#{hit[:required]}",
```
Comment (joshua-ostrom) [inline lib/sarif/semgrep_sarif.rb:41]: I wonder if we need the forbidden and required fields... that may be additional context we can drop here.  Thoughts?

File: lib/sarif/semgrep_sarif.rb [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      details << "\nRequired:#{hit[:required]}" if hit[:required]
+      {
+        id: id(hit[:forbidden], hit[:required]),
+        name: "#{hit[:pattern]} / #{hit[:msg]}",
+        level: "HIGH",
+        details: "#{details}\nHit: #{hit[:hit]}",
+        start_line: location[1],
+        start_column: 1,
+        uri: location[0],
+        help_url: "https://semgrep.dev/docs/writing-rules/rule-syntax/",
```
Comment (ghbren) [inline lib/sarif/semgrep_sarif.rb:91]: Please update all occurrences of the help url to https://github.com/coinbase/salus/blob/master/docs/scanners/semgrep.md

File: spec/lib/sarif/semgrep_sarif_spec.rb [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        repo = Salus::Repo.new("spec/fixtures/semgrep")
+        scanner = Salus::Scanners::Semgrep.new(repository: repo, config: config)
+        scanner.run
+
+        report = Salus::Report.new(project_name: "Neon Genesis")
+        report.add_scan_report(scanner.report, required: true)
+        sarif_report = JSON.parse(report.to_sarif)
+        result = sarif_report["runs"][0]["results"][1]
+        rules = sarif_report["runs"][0]["tool"]["driver"]["rules"]
+        semgrep_doc_url = 'https://github.com/coinbase/salus/blob/master/docs/scanners/semgrep.md'
```
Comment (joshua-ostrom) [inline spec/lib/sarif/semgrep_sarif_spec.rb:73]: Can we use SEMGREP_URI here?

File: lib/sarif/semgrep_sarif.rb [hunk 5]
```diff
@@ -0,0 +1,121 @@
+require 'salus/bugsnag'
+
+module Sarif
+  class SemgrepSarif < BaseSarif
+    include Salus::SalusBugsnag
+
+    SEMGREP_URI = 'https://semgrep.dev/'.freeze
```
Comment (joshua-ostrom) [inline lib/sarif/semgrep_sarif.rb:7]: Just realized we have SEMGREP_URI in here twice.  Safe to drop the first one here?

**Review by joshua-ostrom — APPROVED** [review]

**Pull request merged**